REST API for annotations (back-end comments, take 2)

jaswrks opened the pull request:
## Description

Alternative approaches explored: #4385 and #4386.

A WP REST API endpoint for annotations (back-end comments). Implemented as a custom comment type. The API uses the W3C annotation data model for [annotation selectors](https://www.w3.org/TR/annotation-model/#selectors) and maintains some parity with the W3C annotation object model and protocol — while still doing things the WordPress way, which maximizes compatibility with WP REST API utilities, including those in JavaScript.

### Overview of Changes

#### Introduces:

- `lib/class-wp-annotation-caps.php` to deal with annotation capabilities.
- `lib/class-wp-rest-annotations-controller.php` is the REST API controller.
- `lib/class-wp-annotation-utils.php` are a few annotation utilities.

#### Modifies:

- `lib/load.php`
- `lib/register.php`

#### Near 100% coverage by unit tests

- see: `phpunit/annotations/*`

## API Documentation

**See:** 📄  [https://speca.io/jaswrks/wp-annotations](https://speca.io/jaswrks/wp-annotations)

## How can you test the WP REST API for Annotations?

- Install the [Basic Auth](https://github.com/WP-API/Basic-Auth) plugin for the WP REST API. This allows you to run tests with only a username/password via Postman.

  - Or, simply create this directory and file, which forces the current user to `1` while testing. If you go this route, be sure to delete the file after testing.
  `wp-content/mu-plugins/rest-api-auth-bypass.php`
    ```php
      <?php
      add_action( 'rest_api_init', function() {
          wp_set_current_user( 1 ); // Assuming admin is user ID 1.
          // It's worth testing other roles too; e.g., editor, author, contributor, subscriber.
      } );
    ```

- Now, visit [this page](https://documenter.getpostman.com/view/2613712/collection/7TNeqaX) and click "[Run in Postman](https://documenter.getpostman.com/view/2613712/collection/7TNeqaX)". You can then run your own tests locally in [Postman](https://www.getpostman.com/). Either that, or follow along with the [Annotation API docs](https://speca.io/jaswrks/wp-annotations) and run your own tests over HTTP in any other way you'd like.

  [![2018-01-25_16-17-27](https://user-images.githubusercontent.com/1563559/35420660-4573f34a-01eb-11e8-8505-ea52bd6d16e3.png)](https://documenter.getpostman.com/view/2613712/collection/7TNeqaX)


## How Has This Been Tested?
- Postman (as described above)
- Several unit tests are included with this PR.

## Types of changes
- This approach uses a custom comment type instead of a custom post type.
- This PR strips away support for front-end annotations. Front-end annotations [were explored in a prior PR for the REST API](https://github.com/WordPress/gutenberg/pull/4386), but not a concern at this time.

---

This approach requires a comment query filter to keep annotations out of the comments admin area, and away from plugins that might mistake them as being a more typical comment type. In this PR, the following alters `comments_clauses` in core.

```php
/**
 * Filters WP_Comment_Query clauses.
 *
 * @param  array            $pieces Array of comment query clauses.
 * @param  WP_Comment_Query $query  Current WP_Comment_Query instance.
 *
 * @return array                    Array of comment query clauses.
 */
public static function on_comments_clauses( $pieces, $query ) {
	global $wpdb;

	/*
	 * When cache_domain is 'annotations', only then will annotations be returned by WP_Comment_Query.
	 * Otherwise, if cache_domain is not 'annotations', annotation comment types cannot be returned whatsoever.
	 *
	 * The point being that we want to keep annotations out of any normal comment query performed by core,
	 * and also keep them away from comment-related plugins; i.e., annotations will be unexpected by most plugins.
	 * If a plugin *does* want to query annotations, they should set cache_domain to 'annotations'.
	 */
	if ( 'annotations' !== $query->query_vars['cache_domain'] ) {
		$annotation_types = self::$types;

		foreach ( $annotation_types as &$_type ) {
			$_type = $wpdb->prepare( '%s', $_type );
		}
		$pieces['where'] .= $pieces['where'] ? ' AND ' : '';
		$pieces['where'] .= 'comment_type NOT IN (' . implode( ', ', $annotation_types ) . ')';
	}

	return $pieces;
}
```

## Back-End Annotation Permissions

### Simplified Explanation

- If you can `edit_posts`, and you can `edit_post` (this post). Or, _if you're the post author_. Then you can read all, and create, edit, delete your own back-end annotations in this post.

- Administrators and Editors can edit and delete any post, so they can read, edit, and delete any annotation without restriction.

- Subscribers and the public have no access to back-end annotations whatsoever.

![back-end-permissions](https://user-images.githubusercontent.com/1563559/35069028-6765811a-fb85-11e7-9a58-fa90b676ccb0.png)


## TODO

- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.
- [x] Expand unit test coverage.
- [x] Fix bugs found by @chriskmnds. Noted in comments.

## Non-Blocking TODOs

- [x] Refactor `WP_Annotation_Utils::on_map_meta_cap()` — move logic in each 'case' to a separate function, making a somewhat complex routine easier to follow along with.
- [x] Polish unit tests; e.g., use test generators.
- [x] Move capability-related methods to their own class.
- [x] Remove useless comments in caps class.
- [x] Cleanup comments in utils class.
- [x] Add tests that check valid/invalid UTF-8 sequences.
- [x] Add tests that consider unfiltered HTML.
- [x] fix: Check empty content before & after filters.
- [x] Add tests for annotations in trashed posts.
- [x] Add tests against WP_Comment_Query filters.
- [x] Test setting all comment statuses.
- [x] Cover all selector types in unit tests.
- [x] Test cap checks with missing/invalid args.
- [x] Add tests that expand coverage in REST controller.
- [x] Add tests that expand coverage in utils.
- [x] Fill in any remaining test coverage gaps.
- [x] Document code coverage.
  ```text
  @WP.Annotation::WP_Annotation_Caps
    Methods:  71.43% ( 5/ 7)   Lines:  96.58% (113/117)
  @WP.Annotation::WP_Annotation_Utils
    Methods:  66.67% (10/15)   Lines:  97.71% (256/262)
  @WP.REST.Annotations::WP_REST_Annotations_Controller
    Methods:  85.00% (17/20)   Lines:  98.64% (364/369)
  ```
- [x] Try simplifying cap handlers, https://github.com/WordPress/gutenberg/pull/4685#discussion_r165098655
- [x] Document args to cap handlers, https://github.com/WordPress/gutenberg/pull/4685#discussion_r165092469
- [x] Comment on status considerations in cap handlers, https://github.com/WordPress/gutenberg/pull/4685#discussion_r165096973
- [ ] ~~Consider typehinting WP_REST_Request, https://github.com/WordPress/gutenberg/pull/4685#discussion_r165101602~~
- [ ] Replace `@since [version]` with version in docBlocks.

## Client Side Exploration of Selectors

**jsFiddle:** Building and highlighting a W3C Annotation Selector using XPath.
https://jsfiddle.net/jaswrks/1cLaeyfd/

Commit:
Add REST API for back-end comments (annotations)

chriskmnds commented on lib/class-wp-rest-annotations-controller.php:
```diff
@@ -0,0 +1,1065 @@
+<?php
```
Hey Jason, isn't there a way to break down this humongous file into smaller modules? There is so much going on in here, it will be hard to maintain imo. Same as the utils class above. The separation of what is covered by each is not clear to me either. Maybe start by extracting the permission checks for a start, then gradually organise things maybe per CRUD operation., identify reusable patterns, etc. 

This is obviously a big task, and by all means kudos for writing so much code. Just a little bit more organisation will go a long way.

jaswrks replied:
@chriskmnds Thanks for the review 😄 

> lib/class-wp-rest-annotations-controller.php
> break down this humongous file into smaller modules? 

**WP_REST_Annotations_Controller** extends [WP_REST_Comments_Controller](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-comments-controller.php) in core. There's a [nice article in the handbook](https://developer.wordpress.org/rest-api/extending-the-rest-api/controller-classes/) that summarizes the infrastructure.

While I'm not opposed to separating areas of concern, I think it's important to follow the precedent that's been set and to maintain a consistent pattern.

---

> class-wp-annotation-utils.php
> extracting the permission checks for a start

I agree. Pulling the permissions out of there seems like a great idea.

chriskmnds commented on phpunit/class-rest-annotations-controller-test.php:
```diff
@@ -0,0 +1,1244 @@
+<?php
```
I haven't gone through all the tests here, but I am wondering if you have created test cases to handle all the parameter values and types that are passed to the API. For example, do you check the expected return values when a required parameter is missing, when a wrong value is passed in the request, whether errors are communicated correctly, etc? For example, I get back a status `200 OK` with a value of `Unexpected '<'` in the body of the response when I pass the parameters:

```
{
    "post": 1,
    "via": "gutenberg",
    "content": " ",
    "author": "123"
}
```

Is this normal, am I doing something wrong? Obviously `author: "123"` erred there.

Similarly, passing an empty string for content `content: ""` results in a `400 Bad Request`, while if I pass just a space with `content: " "` results in `200 OK`. I can see though in the response that the `rendered` content is null/empty `""`. Why allow that then and not treat it like the former case? etc..

jaswrks replied:
@chriskmnds Awesome work, thank you! 👁  
It's great to have another pair of eyes on this.

> test cases to handle all the parameter values and types that are passed to the API.

Yes, there were *some* like this already, but most had a focus on user permissions. I've been in the process of restructuring the unit tests for this branch and that work is done now. I just pushed the improved unit tests to this PR and I'll post a summary shortly.

---

Both great finds, and both bugs. Nice work!

- [x] 🐛  **"Unexpected '<' in the body of the response"**... Triggered by an invalid data type in the `author` parameter. This was resulting in an internal `WP_Error` response that went unhandled, resulting in a parse error and a malformed response instead of the error message and code. Fixed by adding [these three lines](https://github.com/jaswrks/gutenberg/blob/f95b0661f9785aa609317502a19527f75027644b/lib/class-wp-rest-annotations-controller.php#L373-L375) and unit tests [now cover this](https://github.com/WordPress/gutenberg/pull/4685/commits/af3b4c1b9de87b3b2edc38b2a0cd28134c0d9333).

- [x] 🐛  **"if I pass just a space with content: " " results in 200 OK."** ... Yeah, I think it should be smarter than this. The same behavior seems to exist in WordPress core. I corrected this for annotations by adding a [smarter analysis](https://github.com/WordPress/gutenberg/pull/4685/commits/f95b0661f9785aa609317502a19527f75027644b#diff-df5e27f1866a42a0c24ae7c0b17b1bc6R1291) of the content and [unit tests](https://github.com/jaswrks/gutenberg/blob/af3b4c1b9de87b3b2edc38b2a0cd28134c0d9333/phpunit/annotations/class-rest-annotations-controller-test.php#L884-L889) also.

jaswrks replied:
Thanks again Chris. I really appreciate you taking the time to look over this with me. Anything I can help you with just hit me on Slack! :-)

jaswrks commented:
Unit tests for this API have been [restructured](https://github.com/jaswrks/gutenberg/tree/add/rest-api-backend-comment-annotations-w3c/phpunit/annotations) to improve their organization, readability, and to  expand coverage. Tests now run approx 30k assertions, with new tests having been added to check things like invalid parameter responses, max string lengths, and more CRUD permutations.

### Two Sets of Tests

- [`class-annotations-test.php`](https://github.com/jaswrks/gutenberg/blob/add/rest-api-backend-comment-annotations-w3c/phpunit/annotations/class-annotations-test.php) Main objective is to test annotation capabilities.
- [`class-rest-annotations-controller-test.php`](https://github.com/jaswrks/gutenberg/blob/add/rest-api-backend-comment-annotations-w3c/phpunit/annotations/class-rest-annotations-controller-test.php) Tests the REST API controller.

jaswrks commented:
### :bug: Bug Fixes

  - Author ID was not being compared as an array in the REST API controller when defining the [`$is_current_user_query`](https://github.com/WordPress/gutenberg/pull/4685/commits/f95b0661f9785aa609317502a19527f75027644b#diff-5b745e0171ea78a848261143d4fd44f0R796) variable. Making it impossible to filter down to only your own annotations. Fixed in [this line](https://github.com/WordPress/gutenberg/pull/4685/commits/f95b0661f9785aa609317502a19527f75027644b#diff-5b745e0171ea78a848261143d4fd44f0R796) and now [covered](https://github.com/jaswrks/gutenberg/blob/f95b0661f9785aa609317502a19527f75027644b/phpunit/annotations/class-rest-annotations-controller-test.php#L489) by unit tests.

  - A whitespace-only `content` string was allowed to pass through when creating or updating a comment. Props @chriskmnds, see: https://github.com/WordPress/gutenberg/pull/4685#discussion_r164389852

  - Submitting a POST request to create a new annotation by a nonexistent `author` resulted in an error internally, but with a broken response. Props @chriskmnds, see: https://github.com/WordPress/gutenberg/pull/4685#discussion_r164389852

  - Trashed annotations having a `_wp_trash_meta_status` that would restore it as public, were wrongly considered readable by other authors in some cases. Fixed in lines like [this one](https://github.com/WordPress/gutenberg/commit/f95b0661f9785aa609317502a19527f75027644b#diff-df5e27f1866a42a0c24ae7c0b17b1bc6R254) and [tested](https://github.com/jaswrks/gutenberg/blob/f95b0661f9785aa609317502a19527f75027644b/phpunit/annotations/class-annotations-test.php#L143-L155).

  - A foreach was [missing its `$prop`](https://github.com/WordPress/gutenberg/pull/4685/commits/f95b0661f9785aa609317502a19527f75027644b?diff=split#diff-5b745e0171ea78a848261143d4fd44f0L947) (and instead, inheriting `$prop` from a previous loop). This was causing wrong and inconsistent behavior in some tests. Fixed and now [covered](https://github.com/jaswrks/gutenberg/blob/f95b0661f9785aa609317502a19527f75027644b/phpunit/annotations/class-rest-annotations-controller-test.php#L765) by unit tests.

  - W3C Selectors were checked by a yoda condition in the [wrong direction](https://github.com/WordPress/gutenberg/pull/4685/commits/f95b0661f9785aa609317502a19527f75027644b?diff=split#diff-df5e27f1866a42a0c24ae7c0b17b1bc6L1130), which caused some valid selectors to fail validation when they were actually fine. Now [covered](https://github.com/jaswrks/gutenberg/blob/f95b0661f9785aa609317502a19527f75027644b/phpunit/annotations/class-rest-annotations-controller-test.php#L810) by unit tests.

  - The existing [`Admin_Test`](https://github.com/WordPress/gutenberg/pull/4685/commits/f95b0661f9785aa609317502a19527f75027644b#diff-8377d145aceec85da6a5cd7c755439aa) was broken when running alongside those for annotations. The `wpSetUpBeforeClass` method should be used instead of `setUpBeforeClass`. I fixed it in this branch so tests will run properly, but I'm going to open a separate PR.

jaswrks commented:
Freshly rebased on master branch and I updated links to commits in my earlier comments.

atimmer reviewed (changes requested):
I've looked at the first part of the code changes and had these remarks.

My main feeling when I look at this PR is that its a ton of changes all at once. I think it could be a lot cleaner if WordPress had first-class support for comment types and private comments.

I am going to look into making this possible in core.

jaswrks commented on lib/class-wp-annotation-caps.php:
```diff
@@ -0,0 +1,441 @@
+	 * @param array  $caps    Required capabilities.
+	 * @param string $cap     Capability to check/map.
+	 * @param int    $user_id User ID (empty if not logged in).
+	 * @param array  $args    Arguments to {@see map_meta_cap()}.
```
Oh, yes. That's a good idea, ty.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,818 @@
+	 * @see WP_Annotation_Utils::on_comments_clauses()
+	 * @see WP_REST_Comments_Controller::get_items()
+	 */
+	public static function on_rest_comment_query( $query_vars, $request ) {
```
Cool, yes. I didn't realize class hints went all the way back to 5.0.

jaswrks replied:
I'm totally in favor of type hints anywhere they're supported, on just about everything. And I got excited when you mentioned that some work all the way back to 5.x. Woohoo! However, since this is not a practice that WordPress has adopted, and there's no mention of hints in the [coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/), I'm going to hold off for now and ask around. Maybe it's fine, I just haven't seen them used in core, have you?

jaswrks replied:
I asked about this on Slack and the impression I have is that it's probably OK to use type hints, but there needs to be some further discussion and a precedent set. Ideally, there would be something added to the WP code standards on this topic before we start using them.

ntwb replied:
Don't worry about creating that Trac ticket @jaswrks I mentioned, this PR will suffice for said discussion 💯

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,818 @@
+
+		$meta_queries = array();
+
+		$vias = $request['via'];
```
For the sake of consistency, array access matches controllers in core; e.g., https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-comments-controller.php

jaswrks commented on lib/class-wp-annotation-caps.php:
```diff
@@ -0,0 +1,441 @@
+/**
+ * Annotations API: WP_Annotation_Caps class
+ *
+ * @package gutenberg
```
A capital seems right to me too. The rest of the PHP files in Gutenberg use lowercase though, so if this changes it should probably happen in one fell swoop as a separate PR. Do you agree?

atimmer replied:
Ah, you did it for consistency. That makes sense. I would leave it like this then. And indeed if we want to change it, it needs to happen in one rename PR.

jaswrks commented on lib/class-wp-annotation-caps.php:
```diff
@@ -0,0 +1,441 @@
+			$wp_trash_meta_status = $wp_trash_meta_status ? $wp_trash_meta_status : '0';
+			$comment_status       = WP_Annotation_Utils::translate_comment_status( $wp_trash_meta_status );
+		}
+		if ( ( 'trash' === $actual_comment_status && ( ! $user_id || (int) $comment->user_id !== $user_id ) )
```
Agree, thanks.

jaswrks commented on lib/class-wp-annotation-caps.php:
```diff
@@ -0,0 +1,441 @@
+	 *
+	 * @return array          Array of required capabilities.
+	 */
+	public static function on_map_meta_cap( $caps, $cap, $user_id, $args ) {
```
Yeah, singleton vs. static. I don't have a strong personal preference between them. There was nothing that needed instantiation in these, so I went with static as one step above global functions. If there are strong opinions about it I'm happy to change.

jaswrks commented on lib/class-wp-annotation-caps.php:
```diff
@@ -0,0 +1,441 @@
+ *
+ * @since [version]
+ */
+final class WP_Annotation_Caps {
```
Great observation, thank you. I was thinking the same thing when I wrote this and tried to avoid some of that with the `get_comment_info()` and `get_post_info()` utils. But the extraction of those variables are now in duplicate and I'd like to improve. Really want to avoid the use of `extract()`. I'll dig in and see if there's anything I can do.

jaswrks replied:
I did some work on this in recent commits. Thanks!

jaswrks commented:
Props @atimmer and @chriskmnds for your reviews. I'm attaching a summary of recent improvements following your suggestions. If you have time to take another look and find any other issues please let know. Thank you both 😄 

### Summary of recent commits:

- Meta-cap handlers moved to their own class; props to @chriskmnds.
- Unit tests for meta-caps moved to their own class.
- Unit tests reorganized. They now live in `phpunit/annotations/*`.
- Many more unit tests and a coverage analysis (near 100% line coverage).
  ```text
  @WP.Annotation::WP_Annotation_Caps
    Methods:  71.43% ( 5/ 7)   Lines:  96.58% (113/117)
  @WP.Annotation::WP_Annotation_Utils
    Methods:  66.67% (10/15)   Lines:  97.71% (256/262)
  @WP.REST.Annotations::WP_REST_Annotations_Controller
    Methods:  85.00% (17/20)   Lines:  98.64% (364/369)
  ```
- Further simplified the meta-cap handlers; props to @atimmer.
- Found one more bug. It was related to empty content checks whenever a user without the `unfiltered_html` cap created or updated an annotation. It was possible to submit a comment containing HTML tags not in the whitelist (which get stripped away) potentially leaving empty content after content filters were applied. Fixed in https://github.com/WordPress/gutenberg/pull/4685/commits/c62eb64b92221bc95493291ae3bf2523303be202.
- I left inline comments on some of the more complex conditionals.
- I docBlock'd the `$args` to each meta-cap handler; props to @atimmer.
- Updated leading topic in this PR with an overview of files it contains, and there's a checklist of my recent work up there also if it helps.

jaswrks commented on lib/class-wp-annotation-caps.php:
```diff
@@ -0,0 +1,441 @@
+	 * @param array  $caps    Required capabilities.
+	 * @param string $cap     Capability to check/map.
+	 * @param int    $user_id User ID (empty if not logged in).
+	 * @param array  $args    Arguments to {@see map_meta_cap()}.
```
Since this function forks itself out to the other cap-specific handlers, I documented the arguments separately in each one to help clarify. Thanks again!

jaswrks commented:
Referencing a trac ticket I was previously unaware of:
[40031: Consider Adding Web Annotations to WordPress](https://core.trac.wordpress.org/ticket/40031)

I invited others from that discussion to leave feedback here and I'm looking forward to other opinions and ideas. My comment here: https://core.trac.wordpress.org/ticket/40031#comment:24

jaswrks commented:
@atimmer writes...
> My main feeling when I look at this PR is that its a ton of changes all at once. I think it could be a lot cleaner if WordPress had first-class support for comment types and private comments.

I agree. As some of you may already know, I went back and forth on this and even completed a large chunk of work in two previous iterations of this API. My first attempt was with a custom post type, then with both front-end _and_ back-end annotations as two different comment types.

I left my thoughts and a list of pros/cons in each of those previous iterations.

- [First attempt with custom post types](https://github.com/WordPress/gutenberg/pull/4385)
- [Second attempt with custom comment types](https://github.com/WordPress/gutenberg/pull/4386)

Then I honed in on private backend annotations only, using **this PR** with a single custom comment type. Only private backend annotations right now, because that's the current Gutenberg objective. But also because I concluded a few things from my experimentation.

#### Front-End Annotation Conclusions

If front-end annotations ever do become a thing, I think they'd be better implemented in a way that would simply extend the existing default `comment` type. In other words, WP core comments could simply be enhanced; allowing them to have annotation-like properties. We don't even need a custom comment type for this, it could just be that comments are allowed to have Selectors like those implemented by this PR, which is what makes a comment an annotation.

That way front-end comments and front-end annotations are both the same thing. The only real 
 difference being that an annotation has a selector; i.e., annotates a specific part of the document. If it doesn't (or is later orphaned) it simply joins or rejoins the flow of comments and is therefore treated like any other front-end comment, because it doesn't select anything.

#### Back-End Annotation Conclusions

What would have made this work *1000%* easier, lighter, and more tightly integrated with core, is if custom comment types were a first-class citizen in core. If core had a `register_comment_type()` and `register_comment_status()` system in place, much of the code for this API and the capabilities associated with it could have been avoided. Alas, we're not there yet, and until we are, the API introduced by this PR seems like the best option to me.

#### My hope is that this work will:

- Satisfy requirements for this task I was given in the best way possible at this time.
- Serve as inspiration to myself or others when considering annotations in core.
- Demonstrate the need for more work to be done on custom comment types in core.

jaswrks commented:
Referencing existing trac ticket for custom comment types: https://core.trac.wordpress.org/ticket/35214

atimmer commented on phpunit/annotations/class-annotation-caps-test.php:
```diff
@@ -0,0 +1,321 @@
+	 *
+	 * @return array Test arrays.
+	 */
+	protected function generate_standard_meta_cap_tests() {
```
Do you know about [`@dataProvider`](https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers)? This makes PHPUnit aware of the data set and tests will continue if one of the data points fails.

jaswrks replied:
Thank you. I experimented with data providers a bit in the unit tests for this API. One thing I have against them is that they slow down unit tests that cover 1000's of permutations. Slower, because each dataset delivered by a data provider results in `setUp()` and `tearDown()` being called on each of those iterations; e.g.,  45k permutations x 2 setup/teardown calls.

I found that tests ran much faster if they were generated in logical sets, with groups of permutations being run together. Resulting in just one `setUp()` and `tearDown()`, which is all that was needed in most of the tests for this API.

Another thing I noticed is that all data providers run before any testing begins. So if a data-provider depends on any database-driven objects, those are lost when tests run, given all the heavy lifting done in `WP_UnitTestCase`.

atimmer replied:
Ah, then it is not feasible.

atimmer reviewed (changes requested):
I have done a more thorough review of the PR. The main changes I would like to see are:

* The annotation status should be in comment meta. I think saving the status in `comment_approved` is an unwise decision.
* Annotations should have an `annotation_type` field (not mentioned in other code review comments). The default should be "comment". Grammatical suggestions or SEO suggestions are of a different subtype. We (@omarreiss and I) think we should have built-in support for different types of annotations from the start.

jaswrks commented:
> I have done a more thorough review of the PR.

@atimmer Thanks for digging in with me here. Your review is much appreciated and I'll take a look at each of the items you pointed out shortly. 😄

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+	 *
+	 * @see wp_get_comment_status()
+	 */
+	public static function get_comment_status( $comment, $check_meta = false ) {
```
If you have time to elaborate just a bit on why you feel this would cause more trouble than it's worth, I'd be curious to know what your experience has been; i.e., more about why you feel that way about `comment_approved`.

> I think these statuses should be put in comment meta. This also makes it easier to migrate to the main comments table if it ever gets a comment_status field.

Isn't that what the `comment_approved` column is already? I realize the name of this field is not ideal, but this is what the family of comment status handlers uses now; e.g., [`wp_set_comment_status()`](https://developer.wordpress.org/reference/functions/wp_set_comment_status/)

atimmer replied:
Ok, you've changed my mind. It is fine to use it like this. Core needs first-class support for this in the future. But that could be done during merge proposal time I think.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+		global $wpdb;
+
+		if ( ! $comment ) {
+			if ( $error ) {
```
I see what you mean. I'm not sure I would have written it this way either, were it not already [like this in `wp_set_comment_status()`](https://github.com/WordPress/WordPress/blob/b9944def27bc8128d7d90f2d22ba915a3d959677/wp-includes/comment.php#L2192-L2193). My intent was mainly to model this function after `wp_set_comment_status()` and add support for custom statuses.

That said, in the end, that `$error` parameter has not been needed by any code in this PR. So I don't see any reason to keep this either.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+		if ( 'annotations' !== $query->query_vars['cache_domain'] ) {
+			$annotation_types = self::$types;
+
+			foreach ( $annotation_types as &$_type ) {
```
I see. Well, this doesn't _need_ to be a reference, but it's a little more concise than the alternative, and I would argue it's easier to read as well. Maybe it's just me though. If you don't agree, I'm happy to change it.

```php
foreach ( $annotation_types as &$_type ) {
    $_type = $wpdb->prepare( '%s', $_type );
}
```
vs.

```php
foreach ( $annotation_types as $_key => $_type ) {
    $annotation_types[ $_key ] = $wpdb->prepare( '%s', $_type );
}
```

atimmer replied:
Ah, right. I didn't see this immediately. I do think the second variation is more readable, even though it is longer.

jaswrks replied:
OK. I'm going to change this, because I can see your point there. TY

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+			 * @param int    Max annotation selector size (in bytes).
+			 * @param array  An array of all selector details.
+			 */
+			$max_selector_size = apply_filters( 'annotation_max_selector_size', $max_selector_size, $selector );
```
No, the [spec](https://www.w3.org/TR/annotation-model/#selectors) doesn't call for a maximum, but it probably should, because selectors can be nested using [refinements](https://www.w3.org/TR/annotation-model/#refinement-of-selection). Some selector types are allowed to contain text, etc. So the goal here is to set an upper limit on size in addition to type-specific validation that occurs below.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+	 *
+	 * @link https://www.w3.org/TR/annotation-model/#selectors
+	 */
+	public static function is_valid_selector( $selector, $recursive = false ) {
```
Great idea! That will be very nice.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+			}
+		}
+
+		switch ( $selector['type'] ) {
```
I remember taking a shot at this myself, but I failed. Each type of selector is allowed to have different properties. Some are allowed to contain refinements, while others cannot. So a JSON schema would need to deal with that logic and support recursive definitions via selector refinements. Ultimately, I felt the JSON schema should simply define the selector types. We can handle the logic and recursive validation here.

atimmer replied:
Hmm, maybe out of scope for this PR, but I really like the Symfony Config component: https://symfony.com/doc/current/components/config/definition.html#variable-nodes. It is a readable and semantic way to define a nested configuration structure.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+					'conformsTo',
+					'refinedBy',
+				);
+				if ( array_diff_key( $selector, array_fill_keys( $allow_keys, 0 ) ) ) {
```
Agree, thanks.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+		}
+
+		// @codeCoverageIgnoreStart
+		// Testing would require a DB write failure.
```
Good points, thank you. There were three or four sections like this I was unable to cover. I'll start by just removing those comments and then reassess once the next iteration is ready for review.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+	 * @return bool           True if empty, if it's not a string, or the blog's charset
+	 *                        is UTF-8 and it's invalid UTF-8.
+	 */
+	public static function is_content_empty( $content ) {
```
One could argue this function is over-the-top. On the other hand, without this level of detail it would be possible to submit something like the example you gave `<span></span>`. Strictly speaking, that's not empty, but it might as well be, because it produces an empty comment.

Another example would be a user who lacks the `unfiltered_html` capability submitting an annotation containing only tags they're not allowed to use, which are stripped away when the content is filtered, leaving an empty string, or empty-ish remnants; e.g, `<span></span>`.

So the point here is to do the best we can at enforcing this rule; i.e., that every annotation must be accompanied by something meaningful, and where `<span></span>` is not considered meaningful, in the same way `''` is not considered meaningful.

jaswrks replied:
Can you think of a reason why `<span></span>` should be allowed through server-side?

atimmer replied:
I don't feel that strongly about this. The only thing is that it's pretty complex validation logic for arguably little gain.

jaswrks commented:
@atimmer Very helpful review. I'll go over the rest a bit later. Thanks again!

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+		if ( $meta_queries ) {
+			if ( ! empty( $query_vars['meta_query'] ) ) {
+				// @codeCoverageIgnoreStart
+				// Testing would require a plugin modifying query.
```
https://github.com/WordPress/gutenberg/pull/4685#discussion_r166947938

jaswrks commented on lib/class-wp-rest-annotations-controller.php:
```diff
@@ -0,0 +1,1097 @@
+		$comment = get_comment( $comment_id );
+
+		// @codeCoverageIgnoreStart
+		// Testing would require a cache or DB read failure.
```
https://github.com/WordPress/gutenberg/pull/4685#discussion_r166947938

jaswrks commented:
@atimmer @omarreiss

What would you like to shoot for in support of multiple annotation types?

I see two workable solutions:

1. Multiple annotation comment types. For example, `annotation`, `seo_annotation`, etc. This is how it works now and this approach could be improved by adding better filters.

2. One `annotation` comment type, and then something like the `annotation_type` meta that you suggested. I assume it is this approach you like best, correct?

Can you think of any pros/cons associated with each of those?

One reason I went with option **1** is because I felt it would be easier to merge into a future WordPress core that supports `register_comment_type()`. When/if that occurs, then using multiple comment types may allow greater flexibility; e.g., each type of annotation could be associated with different capabilities and other type-specific settings.

jaswrks commented on lib/class-wp-rest-annotations-controller.php:
```diff
@@ -0,0 +1,1097 @@
+				$is_current_user_query = $current_user_id && array( $current_user_id ) === $request['author'];
+
+				foreach ( array(
+					'author',
```
The intention here is to prevent non-administrative users (e.g., authors/contributors) from looking at another user's history. The `author`, `author_exclude`, and `author_email` parameters are also guarded in core, [in a similar way, for comments](https://github.com/WordPress/WordPress/blob/b9944def27bc8128d7d90f2d22ba915a3d959677/wp-includes/rest-api/endpoints/class-wp-rest-comments-controller.php#L138). A useful exception to this rule is that any user who can edit, can list their own annotations; i.e., note the `$is_current_user_query` variable a few lines below this.

atimmer replied:
That makes sense. I think it warrants an inline comment.

jaswrks commented on lib/class-wp-rest-annotations-controller.php:
```diff
@@ -0,0 +1,1097 @@
+		// @codeCoverageIgnoreStart
+		// Not possible to test at this time.
+		// Strengthens security in custom post types and on sites with permission mods.
+		if ( ! current_user_can( 'edit_annotations', $type ) ) {
```
Great suggestion. I'll elaborate on this in the comments. The rationale for this is not great at the moment, as it is currently impossible to `create_annotations` if you can't `edit_annotations`.

I left this in only because it would prevent date/meta from being set in the case of a custom post type or plugin altering permissions in such a way that it allows annotation creation, but locks editing. Those two fields would then be restricted in such a case, which seems desirable in that scenario.

I'm going review this again though, because it _is_ very confusing and way out on the edge. I'd prefer to get rid of it in favor of a more straightforward approach.

atimmer replied:
Hmm, but I would argue that you should be allowed to at least set meta on an annotation if you are not allowed to edit it. For example, in the scenario where we would persist annotations from Yoast SEO, we would like to save some extra data in meta. But then users who can only create annotation (not edit) would not be able to set this extra data.

jaswrks replied:
Good point. I have no problem loosening permissions here by removing `meta` from the list.

A scenario where a user can create, but not edit, is not one that would occur unless permissions were altered by a plugin. I will leave a portion of this in to protect the `date` and `date_gmt` fields in that scenario, but I'll remove `meta` as you point out.

jaswrks commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+	 *
+	 * @return stdClass|null           Data object. Null on failure.
+	 */
+	public static function get_comment_data( $comment ) {
```
Good suggestion. I don't see any reason we can't do this, and I agree it would simplify this.

atimmer commented on lib/class-wp-annotation-utils.php:
```diff
@@ -0,0 +1,820 @@
+		 *
+		 * @param array Comment types allowed for annotations.
+		 */
+		$allow_types = apply_filters( 'annotation_allow_types', self::$types );
```
I mean that this filter should be applied in the `on_comments_clauses` method. Otherwise, a type added by this filter will be allowed, but will also be shown in default comment overview. Such as the front-end.

jaswrks replied:
Right, I see what you meant now. That will get fixed when I move the filter and change the name. Definitely agree.

atimmer commented:
@jaswrks The goal of multiple annotation types is to make sure the UI can default to only showing one type of annotation. I think that is possible using different annotation types. I'll discuss this with @omarreiss again to see if that achieves the goals we want to hit.

atimmer commented:
I've discussed this and we think that having different comment types is 👌🏻.

jaswrks commented on lib/class-wp-rest-annotations-controller.php:
```diff
@@ -0,0 +1,1097 @@
+
+		if ( ! $post_ids ) {
+			return new WP_Error( 'rest_missing_annotation_posts', __( 'Missing post(s).', 'gutenberg' ), array(
+				'status' => 400,
```
I agree with `400` error codes being out of place here.

The REST API is rather interesting in its approach to permissions. To properly check permissions one must take the current `WP_REST_Request` into consideration, which is a variable that's passed into these permission checks.

The only problem with this is that [`rest_send_allow_header()`](https://developer.wordpress.org/reference/functions/rest_send_allow_header/) filters every API response, and it calls each of the permission checks with the current `WP_REST_Request`, which may or may not contain the type of request parameters expected; e.g., it may call `get_items_permissions_check( $request )` with request parameters originally intended to create a new annotation.

That makes it more difficult to write a permission check that considers the current request, because at times the current request contains, or doesn't contain, all the information necessary to check permissions properly.

There might be something I can do to improve this though.

jaswrks commented:
👁👁   I could use another pair of eyes on this now. I think I've covered all recent requests and enhancements. I've added two new classes, some new functions, and new filters to make it easier for plugins to extend annotations.

- New class: [`WP_Annotation`](https://github.com/jaswrks/gutenberg/blob/1bac8be7e92daa1c0a932b68f8dbd954916f8920/lib/class-wp-annotation.php#L14)
- New class: [`WP_Annotation_Selector`](https://github.com/jaswrks/gutenberg/blob/1bac8be7e92daa1c0a932b68f8dbd954916f8920/lib/class-wp-annotation-selector.php#L26)
- New function and filter in: [`get_annotation()`](https://github.com/jaswrks/gutenberg/blob/1bac8be7e92daa1c0a932b68f8dbd954916f8920/lib/annotations.php#L21)
- New function and filter in: [`get_annotation_types()`](https://github.com/jaswrks/gutenberg/blob/1bac8be7e92daa1c0a932b68f8dbd954916f8920/lib/annotations.php#L63)
- New function and filter in: [`get_default_annotation_type()`](https://github.com/jaswrks/gutenberg/blob/1bac8be7e92daa1c0a932b68f8dbd954916f8920/lib/annotations.php#L85)
- New function and filter in: [`get_annotation_statuses()`](https://github.com/jaswrks/gutenberg/blob/1bac8be7e92daa1c0a932b68f8dbd954916f8920/lib/annotations.php#L110)
- New function and filter in: [`get_default_annotation_status()`](https://github.com/jaswrks/gutenberg/blob/1bac8be7e92daa1c0a932b68f8dbd954916f8920/lib/annotations.php#L155)
- Removed three or four instances of `@codeCoverageIgnoreStart`.
- Removed an iteration by reference in favor of clarity.
- Added inline comments to explain complex conditionals.
- Removed `meta` from restricted fields when creating
  as a user who lacks the ability to edit annotations.
- Improved status codes in REST API controller permission checks.
- Updated all existing unit tests following these changes.
- Added new unit tests for the `WP_Annotation` class.

### Getting an Annotation Instance

```php
<?php
$annotation = get_annotation( 123 ); // WP_Annotation
$selector   = $annotation->get_selector(); // WP_Annotation_Selector
echo $selector->type; // e.g., XPathSelector
```

### How to Add Annotation Types

```php
<?php
add_filter( 'get_annotation_types', function( $types ) {
    return array_merge( $types, array( 'seo_annotation' ) );
} );
```
### How to Add Annotation Statuses

```php
<?php
add_filter( 'get_annotation_statuses', function( $statuses ) {
    return array_merge( $statuses, array( 'accepted' ) );
} );
```

aduth commented:
Is the work here still relevant, and can it be reasonably be brought up to date with current master? Or can we close it?

atimmer commented:
The work is still relevant, however, I haven't prioritized this part. That's because a full-fledged commenting API with persistence is a ton more work than I have the capacity for. So I have focused on getting client side annotations to work because it will give us a feature that will work right now. Client-side annotation is also required for feature parity of the Yoast plugin with the classic editor.

It's good to note that I am keeping in mind this PR, so all my work is future compatible with the commenting API.

gziolo commented:
@atimmer should we move it to Trac?

atimmer commented:
@gziolo I don't see us working on this in the very short term, so I don't know if there is a lot of value of moving this to trac now.

gziolo commented:
It seems like it isn't actionable in the upcoming months. It wasn't listed on the roadmap for 2019: https://make.wordpress.org/core/2018/12/08/9-priorities-for-2019/. Collaborative editing is planned for the 3rd phase of Gutenberg development. Let's close this PR and get back to it later. We can always reopen 😄 

@jaswrks many thanks for your work done so far. I hope we can reuse most of the code in the future 🙇
